rts/at/integration/src/main/java/org/jboss/narayana/rest/integration/RecoveryManager.java

```diff
@@ -93,12 +101,15 @@ public void removeParticipantInformation(final ParticipantInformation participan
         }
 
         final RecoveryStore recoveryStore = StoreManager.getRecoveryStore();
-        final Uid uid = new Uid(participantInformation.getId());
+        String participantId = participantInformation.getId();
 
-        try {
-            recoveryStore.remove_committed(uid, PARTICIPANT_INFORMATION_RECORD_TYPE);
-        } catch (ObjectStoreException e) {
-            LOG.warn("Failure while removing participant information from the object store.", e);
+        if(committedParticipants.get(participantId) != null) {
+            try {
+                recoveryStore.remove_committed(new Uid(participantId), PARTICIPANT_INFORMATION_RECORD_TYPE);
+                committedParticipants.remove(participantId);
+            } catch (ObjectStoreException e) {
+                LOG.warn("Failure while removing participant information from the object store.", e);
```

Comment:
I think the record should not be removed from the map if the `remove_committed` fails. It means that the object store was not updated. It's important to ensure to really cleaning the object store in the next attempt. The warning sounds correct here as it's needed to warn the user that the participant could not be removed because of some internal failure, probably on the storage.

Reply:
The record will not be removed. However subsequent attempts to remove it will not be attempted (if we failed once then it is likely to continually fail). The way you have it coded is that if the remove fails the warning will appear in the log on every recovery pass. A single WARNING in the log should be sufficient for notifying the user about this error.

Reply:
I really don't agree on this strategy. When a record is not removed from the log then it's the Narayana strategy to inform the user about the consistent failure.
My idea is to really let the warning being printed on every recovery pass.

Reply:
Normally we only continually report recovery failures when we are expecting recovery to happen and repeating warnings for that purpose is acceptable. But failing to delete file will most likely fail forevermore and the user only needs to be told about it once.

Reply:
@mmusgrov - is it not possible to have a transient storage problem though? Perhaps we need to understand better _why_ the remove is failing. But that might be better in some object store API change, like a return type that determines if the failure is expected to be transient or not. 

Your proposed change (in the first comment of this thread) does not look to be rendering fully to me (or Ondra already took that in). How do you propose to suppress subsequent warnings? I am not aware of a precedent to do that in Narayana so far. Doesn't the recovery manager report repeated failures to recover things? But if there are other examples of that let's evaluate as repeating the message at WARN at least might not be so useful (perhaps trace) and might be covered by wording the message that the failure will not be re-reported?

BTW, given the current state of the patch (in that we only put it in the map when write_committed passes or it is recovered/recreated, I can't guess why remove_committed would throw an exception now but maybe there is something in the recover/recreated paths that would do that?

Reply:
@mmusgrov I don't agree that user should be informed about the failure of the storage just once. When the storage is not cleaned the e.g. rollback operation will be retried. Probably with some error. Additional information that there is the trouble with removal is important part of the context.

@tomjenkinson  the `remove_committed` is not expected to throw an exception in the current state of the patch. It just logs a warning that there is not a way to remove the record.

Reply:
@ochaloup I guess Mike was trying to prevent was some outcome a bit like the following (and this is total pseudo log statements and pseudo code):

```
while true {
  // rest of the algorithm, then
  "Failure while removing participant information from the object store" with a caused by of some permanent error like "File does not exist"
}
```

What I am not sure of whether that could ever be the case. And if it could, rather than suppressing the logging alone it would perhaps be better to consider what can cause the condition and deal with it in the algorithm to give a break out clause?

Reply:
> @tomjenkinson the `remove_committed` is not expected to throw an exception in the current state of the patch. It just logs a warning that there is not a way to remove the record.

Responding to this separate. I was not clear enough, let me rephrase: given the current state of the patch, I can't guess why there would be problem to remove the record but maybe there is something in the recover/recreated paths that would cause there to be a problem to remove it somehow?

Reply:
@tomjenkinson ok. maybe I understand, at least partially.

@tomjenkinson @mmusgrov  First, for the understanding of the changes, please, consult the source code. Where the method is called from and what is the context it's called from.
It's quite possible I misunderstand something but if we are talking from start in some semi-real context it's something which I'm not able to follow and which makes my work pretty hard.

@tomjenkinson the failure of the removal is the same from any part of the code that uses the object store. There is some IO failure or some internal storage failure that makes impossible to remove the record. It's something which can't be handled by algorithm. And in case there is some adjustment possible then it has to be handled by the `ObjectStore` API in arjunacore not here in the `rts/at` module.

Reply:
Thanks for clarifying @ochaloup. If the patch introduces no likelihood of causing the remove_committed to fail then I think changes to dealing with non-recoverable remove_committed errors should be dealt with as an RFE

Reply:
ot possible to have a transient storage problem though? Perhaps we need to understand better _why_ the remove is failing. But that might be better in some object store API change, like a return type that determines if the failure is expected to be transient or not.
> 
> Your proposed change (in the first comment of this thread) does not look to be rendering fully to me (or Ondra already took that in). How do you propose to suppress subsequent warnings? I am not aware of a precedent to do that in Narayana so far. Doesn't the recovery manager report repeated failures to recover things? But if there are other examples of that let's evaluate as repeating the message at WARN at least might not be so useful (perhaps trace) and might be covered by wording the message that the failure will not be re-reported?
> 
> BTW, given the current state of the patch (in that we only put it in the map when write_committed passes or it is recovered/recreated, I can't guess why remove_committed would throw an exception now but maybe there is something in the recover/recreated paths that would do that?

Reply:
I am adding my approval of this PR on the following basis:

- XTS does the same as what's in this PR.
- The error should may only ever occur each time an InboundBridgeManager is created (after a server restart) and after a transaction is finished.
- As far as I can see this new code is not ran by Periodic Recovery so my concern about filling the logs on each recovery pass is not a valid concern.

But I'd still argue that my points made in the previous comment are still valid but not sufficient to block this PR.